src/drawscene.cpp

```diff
@@ -459,21 +459,55 @@ void draw_pageflip_3d_mode(Camera& camera, bool show_hud,
 #endif
 }
 
-void draw_plain(Camera &camera, bool show_hud, Hud &hud,
-		video::IVideoDriver *driver, bool draw_wield_tool,
-		Client &client, gui::IGUIEnvironment *guienv)
+inline int scaledown(int coef, int size)
 {
+	return (size + coef - 1) / coef;
+}
+
+void draw_plain(Camera &camera, bool show_hud,
+		Hud &hud, video::IVideoDriver *driver,
+		scene::ISceneManager *smgr, const v2u32 &screensize,
+		bool draw_wield_tool, Client &client, gui::IGUIEnvironment *guienv,
+		video::SColor skycolor)
+{
+// Undersampling-specific stuff
+	static video::ITexture *image = NULL;
+	static v2u32 last_pixelated_size = v2u32(0, 0);
+	int undersampling = g_settings->getU16("undersampling");
+	v2u32 pixelated_size;
+	v2u32 dest_size;
+	if (undersampling) {
+		pixelated_size = v2u32(scaledown(undersampling, screensize.X),
+				scaledown(undersampling, screensize.Y));
+		dest_size = v2u32(undersampling * pixelated_size.X, undersampling * pixelated_size.Y);
```

Comment:
it shouldn’t add black borders as it rounds upwards. performing integer stretching only is important for good look esp. with no AA.